Implement WebTransportOptions.serverCertificateHashes

achristensen07 opened the pull request:
#### 3fbb5096bc1f6a22ceba56f3759f30afc24e0205
<pre>
Implement WebTransportOptions.serverCertificateHashes
<a href="https://bugs.webkit.org/show_bug.cgi?id=300057">https://bugs.webkit.org/show_bug.cgi?id=300057</a>
<a href="https://rdar.apple.com/160151042">rdar://160151042</a>

Reviewed by Matthew Finkel.

This allows JS to specify the hashes of the certificate used for TLS in WebTransport instead of using PKI.

Tests: Tools/TestWebKitAPI/Tests/WebKitCocoa/WebTransport.mm
       Tools/TestWebKitAPI/WebTransportServer.h
       Tools/TestWebKitAPI/WebTransportServer.mm
* Source/WebKit/NetworkProcess/webtransport/cocoa/NetworkTransportSessionCocoa.mm:
(WebKit::leafCertificateMatchesWebTransportHash):
(WebKit::didReceiveServerTrustChallenge):
(WebKit::createParameters):
(WebKit::NetworkTransportSession::create):
* Tools/TestWebKitAPI/Tests/WebKitCocoa/WebTransport.mm:
(TestWebKitAPI::TEST(WebTransport, ServerCertificateHashes)):
* Tools/TestWebKitAPI/WebTransportServer.h:
* Tools/TestWebKitAPI/WebTransportServer.mm:
(TestWebKitAPI::WebTransportServer::WebTransportServer):

Canonical link: <a href="https://commits.webkit.org/300917@main">https://commits.webkit.org/300917@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/d1459b24b50ab92d6174342c5a3cb6d0a3cf42aa

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows | Apple Internal |
| ----- | ---------------------- | ------- |  ----- |  --------- | ------ |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/124281 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/43967 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/34691 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/131117 "Built successfully") | [  ~~🛠 win~~](https://ews-build.webkit.org/#/builders/59/builds/76350 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 ios-apple](https://ews-bridge.webkit.apple.com/builds/sw/T-276/92699e7b-e2c4-41dd-8b79-0fc7c00d5b52/f7f35391-94c2-4c65-9a0c-885e3429986d) 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/126158 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/44710 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/52569 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/94538 "Passed tests") | [  ~~🧪 win-tests~~](https://ews-build.webkit.org/#/builders/59/builds/76350 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [![loading](https://user-images.githubusercontent.com/3098702/171232313-daa606f1-8fd6-4b0f-a20b-2cb93c43d19b.png) 🛠 mac-apple](https://ews-bridge.webkit.apple.com/builds/sw/T-276/92699e7b-e2c4-41dd-8b79-0fc7c00d5b52/59e80c19-10a8-4a24-8553-268836036149) 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/127235 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/132/builds/35629 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/111165 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/75126 "Passed tests") | | [![loading](https://user-images.githubusercontent.com/3098702/171232313-daa606f1-8fd6-4b0f-a20b-2cb93c43d19b.png) 🛠 vision-apple](https://ews-bridge.webkit.apple.com/builds/sw/T-276/92699e7b-e2c4-41dd-8b79-0fc7c00d5b52/09ceaca1-f148-44b3-8b2a-bb7d9899ae5c) 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/133/builds/34573 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/135/builds/29324 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/74598 "Built successfully") | | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/105373 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/136/builds/29546 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/133788 "Built successfully") | | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/51192 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/122/builds/39037 "Passed tests") | [❌ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/103016 "2 flakes 54 failures") | | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/51585 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/107381 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/102814 "Passed tests") | | 
| | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/48172 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/137/builds/26429 "Passed tests") | [  ~~🛠 playstation~~](https://ews-build.webkit.org/#/builders/134/builds/48132 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/22/builds/19523 "Built successfully and passed tests") | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/51056 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/56838 "Built successfully") | | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/50493 "Built successfully") | | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/53849 "Built successfully") | | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/52168 "Built successfully") | | | | 
<!--EWS-Status-Bubble-End-->

sysrqb commented on Source/WebKit/NetworkProcess/webtransport/cocoa/NetworkTransportSessionCocoa.mm:
```diff
@@ -69,8 +69,60 @@
+        return false;
+
+    std::array<uint8_t, CC_SHA256_DIGEST_LENGTH> sha2;
+    CC_SHA256([x509Data bytes], [x509Data length], sha2.data());
```
nit: `bytes` and `length` are properties, so could use `.`

achristensen07 replied:
I'd need to use .get(). and I think this is a little more concise.

sysrqb commented on Source/WebKit/NetworkProcess/webtransport/cocoa/NetworkTransportSessionCocoa.mm:
```diff
@@ -69,8 +69,60 @@
+    CC_SHA256([x509Data bytes], [x509Data length], sha2.data());
+
+    bool hashMatches = std::ranges::any_of(hashes, [&] (auto& hash) {
+        return hash.algorithm == "sha-256"_s && equalSpans(std::span(sha2), hash.value.span());
```
Does/should the `hash.algorithm` check be case-insensitive? or it must match exactly?

achristensen07 replied:
The spec says yes.

sysrqb commented on Source/WebKit/NetworkProcess/webtransport/cocoa/NetworkTransportSessionCocoa.mm:
```diff
@@ -69,8 +69,60 @@
+        return false;
+
+    CFTimeInterval validityPeriod = endTime - beginTime;
+    constexpr double twoWeeks = 2 * 7 * 24 * 60 * 60;
```
nit: could use something like `2 * 7 * secondsPerDay` (but it's only marginally more readable)

achristensen07 committed:
Implement WebTransportOptions.serverCertificateHashes
https://bugs.webkit.org/show_bug.cgi?id=300057
rdar://160151042

Reviewed by Matthew Finkel.

This allows JS to specify the hashes of the certificate used for TLS in WebTransport instead of using PKI.

Tests: Tools/TestWebKitAPI/Tests/WebKitCocoa/WebTransport.mm
       Tools/TestWebKitAPI/WebTransportServer.h
       Tools/TestWebKitAPI/WebTransportServer.mm
* Source/WebKit/NetworkProcess/webtransport/cocoa/NetworkTransportSessionCocoa.mm:
(WebKit::leafCertificateMatchesWebTransportHash):
(WebKit::didReceiveServerTrustChallenge):
(WebKit::createParameters):
(WebKit::NetworkTransportSession::create):
* Tools/TestWebKitAPI/Tests/WebKitCocoa/WebTransport.mm:
(TestWebKitAPI::TEST(WebTransport, ServerCertificateHashes)):
* Tools/TestWebKitAPI/WebTransportServer.h:
* Tools/TestWebKitAPI/WebTransportServer.mm:
(TestWebKitAPI::WebTransportServer::WebTransportServer):

Canonical link: https://commits.webkit.org/300917@main
